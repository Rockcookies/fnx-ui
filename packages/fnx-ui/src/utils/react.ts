import {
	Children,
	forwardRef,
	ForwardRefExoticComponent,
	ForwardRefRenderFunction,
	FunctionComponent,
	PropsWithoutRef,
	ReactElement,
	RefAttributes,
} from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};

export const createForwardRef = <R, P = any>(
	displayName: string,
	render: ForwardRefRenderFunction<R, P>,
): ForwardRefExoticComponent<PropsWithoutRef<P> & RefAttributes<R>> => {
	const fc = forwardRef(render);
	fc.displayName = displayName;
	return fc;
};

export const createDefaultsForwardRef = <R, P, D>(
	displayName: string,
	defaultProps: D,
	render: ForwardRefRenderFunction<R, P & D>,
): ForwardRefExoticComponent<
	PropsWithoutRef<P & Partial<D>> & RefAttributes<R>
> => {
	const fc = forwardRef(render);
	fc.displayName = displayName;
	fc.defaultProps = defaultProps;
	return fc as any;
};

export const createFC = <P = any>(
	displayName: string,
	fc: FunctionComponent<P>,
): FunctionComponent<P> => {
	fc.displayName = displayName;
	return fc;
};

export const createDefaultsFC = <P, D>(
	displayName: string,
	defaultProps: D,
	fc: FunctionComponent<P & D>,
): FunctionComponent<P & Partial<D>> => {
	fc.displayName = displayName;
	fc.defaultProps = defaultProps;
	return fc as any;
};
