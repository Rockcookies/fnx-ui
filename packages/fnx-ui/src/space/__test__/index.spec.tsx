import React from 'react';
import { render } from '@testing-library/react';
import Space from '../index';
import Button from '../../button';

describe('<Space/>', () => {
	it('should render space', async () => {
		const { container } = render(
			<Space>
				<Button>Button</Button>
				<Button>Button</Button>
				<Button>Button</Button>
			</Space>,
		);

		const items =
			container.querySelectorAll<HTMLDivElement>('.fnx-space-item');

		expect(items[0].style.marginRight).toBe('8px');
		expect(items[1].style.marginRight).toBe('8px');
		expect(items[2].style.marginRight).toBe('');
	});

	it('should render vertical', async () => {
		const { container, getByTestId } = render(
			<Space direction="vertical" fill data-testid="space">
				<Button block>Button</Button>
				<Button block>Button</Button>
				<Button block>Button</Button>
			</Space>,
		);

		const space = getByTestId('space');
		expect(space.classList).toContain('fnx-space--vertical');

		const items =
			container.querySelectorAll<HTMLDivElement>('.fnx-space-item');
		expect(items[0].style.marginBottom).toBe('8px');
		expect(items[1].style.marginBottom).toBe('8px');
		expect(items[2].style.marginBottom).toBe('');
	});

	it('should render empty space', async () => {
		const { container } = render(<Space></Space>);

		expect(container).toMatchSnapshot();
	});

	it('should render size 20px', async () => {
		const { container } = render(
			<Space size={20}>
				<Button block>Button</Button>
				<Button block>Button</Button>
				<Button block>Button</Button>
			</Space>,
		);

		const items =
			container.querySelectorAll<HTMLDivElement>('.fnx-space-item');

		expect(items[0].style.marginRight).toBe('20px');
		expect(items[1].style.marginRight).toBe('20px');
		expect(items[2].style.marginRight).toBe('');
	});

	it('should render size 2rem', async () => {
		const { container } = render(
			<Space size={['2rem']}>
				<Button block>Button</Button>
				<Button block>Button</Button>
				<Button block>Button</Button>
			</Space>,
		);

		const items =
			container.querySelectorAll<HTMLDivElement>('.fnx-space-item');

		expect(items[0].style.marginRight).toBe('2rem');
		expect(items[1].style.marginRight).toBe('2rem');
		expect(items[2].style.marginRight).toBe('');
	});

	test('should render align start', async () => {
		const { getByTestId } = render(
			<Space align="start" data-testid="space">
				<Button block>Button</Button>
				<Button block>Button</Button>
				<Button block>Button</Button>
			</Space>,
		);

		const space = getByTestId('space');
		expect(space.classList).toContain('fnx-space--align-start');
	});

	test('should render wrap', async () => {
		const { getByTestId } = render(
			<Space wrap data-testid="space">
				<Button block>Button</Button>
				<Button block>Button</Button>
				<Button block>Button</Button>
				<Button block>Button</Button>
				<Button block>Button</Button>
				<Button block>Button</Button>
				<Button block>Button</Button>
				<Button block>Button</Button>
				<Button block>Button</Button>
			</Space>,
		);

		const space = getByTestId('space');
		expect(space.classList).toContain('fnx-space--wrap');
	});
});
