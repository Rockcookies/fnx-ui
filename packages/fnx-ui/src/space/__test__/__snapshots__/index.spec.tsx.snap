// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Space/> should render empty space 1`] = `<div />`;
